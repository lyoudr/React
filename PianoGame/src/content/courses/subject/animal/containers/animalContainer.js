import { connect } from 'react-redux';
import Animal from '../presentcomp/Animal';
import {showCatDetail, notshowCatDetail} from '../../../../../redux/actions/index';

const mapStateToProps = state => {
  return ({
    cats: state.animal.cats,
  })
};

const mapDispatchToProps = dispatch => ({
  showCatDetail: (id) => dispatch(showCatDetail(id)),
  notshowCatDetail : (id) => dispatch(notshowCatDetail(id))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Animal);