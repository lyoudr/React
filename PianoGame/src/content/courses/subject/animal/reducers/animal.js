import cat0 from '../../../../../assets/images/cat_0.jpg';
import cat1 from '../../../../../assets/images/cat_1.jpg';
import cat2 from '../../../../../assets/images/cat_2.jpg';
import cat3 from '../../../../../assets/images/cat_3.jpg';

const AnimalData = {
  cats : [
    {
      "name": "Kitty",
      "describe": "Some quick example text to build on the card title and make up the bulk of the card's content.",
      "imgsrc": cat0,
      "index": 0,
      "detail": 'This is Kitty',
      "title": "Persian cat",
      "show" : false,
      "guide": "In the Middle East, region they are widely known as 'Iranian cat' and in Iran they are known as 'Shirazi cat'. The first documented ancestors of the Persian were imported into Italy from Iran (historically known as Persia in the west) around 1620."
    },
    {
      "name": "Cathy",
      "describe": "Some quick example text to build on the card title and make up the bulk of the card's content.",
      "imgsrc": cat1,
      "index": 1,
      "detail": 'This is Cathy',
      "title": "British Shorthair",
      "show" : false,
      "guide": "The British Shorthair is the pedigreed version of the traditional British domestic cat, with a distinctively chunky body, dense coat and broad face. The most familiar color variant is the 'British Blue', a solid blue-gray with copper eyes, medium tail, but the breed has also been developed in a wide range of other colours and patterns, including tabby and colorpoint."
    },
    {
      "name": "KiKi",
      "describe": "Some quick example text to build on the card title and make up the bulk of the card's content.",
      "imgsrc": cat2,
      "index": 2,
      "detail": 'This is KiKi',
      "title": "Bengal cat",
      "show" : false,
      "guide": "The Bengal cat is a domesticated cat breed created from hybrids of domestic cats, the Asian leopard cat (Prionailurus bengalensis) and the Egyptian Mau, which gives them their golden shimmer â€“ the breed name comes from the taxonomic name."
    },
    {
      "name": "WiWi",
      "describe": "Some quick example text to build on the card title and make up the bulk of the card's content.",
      "imgsrc": cat3,
      "index": 3,
      "detail": 'This is WiWi',
      "title": "Abyssinian cat",
      "show": false,
      "guide": "The Abyssinian is a slender, fisne-boned, medium-sized cat. The head is moderately wedge shaped, with a slight break at the muzzle, and nose and chin ideally forming a straight vertical line when viewed in profile. They have alert, relatively large pointed ears."
    },
  ]
}

export const animal = (state = AnimalData, action) => {
  console.log('state is =>', state);
  switch(action.type) {
    case 'SHOW_CATDETAIL':
      const showcats = state.cats.map((cat, index) => {
        if(index === action.id){
          cat.show = true;
        };
        return cat;
      });
      return Object.assign({}, state, {cats: showcats});
    case 'NOTSHOW_CATDETAIL':
      const notshowcats = state.cats.map((cat, index) => {
        if(index === action.id){
          cat.show = false;
        };
        return cat;
      });
      return Object.assign({}, state, {cats: notshowcats});
    default: 
      return state;
  }
} 


/*Notes => Why Redux doesn't re-render component? */
// should put "state" to return statement to notify "mapStateToProps" to tell component to re-render
