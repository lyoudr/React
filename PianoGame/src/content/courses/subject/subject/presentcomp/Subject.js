import React, { useState, useEffect } from 'react';
import { EachSubjectData } from '../../Subjectlist';
import defaultImg from '../../../../../assets/images/subject_default.jpeg';
import '../../../../../assets/sass/courses/subject/subject.scss';
import jQuery from 'jquery';

/* Hooks */

const AddNotetoTextFunc = React.forwardRef((props, ref) => {
  const textRef = React.createRef();

  const [childnotes, setNotes] = useState([]);
  useEffect(() => {
    if (props.notes && props.elemId) {
      let foundnote = props.notes.find(item => {
        return item.id === props.elemId;
      });
      setNotes(foundnote.textNote);
    }
  });

  const addNote = () => {
    let newnote = {};
    newnote[props.selectedText] = textRef.current.value;
    setNotes(oldarr => [...oldarr, newnote]);
    props.hintRef.current.classList.remove('active');
    textRef.current.value = '';
  }

  return (
    <div className="addnotetoText" ref={ref}>
      <textarea ref={textRef}></textarea>
      <NoteContext.Consumer>
        {({ addNotetoText }) => (
          <div className="submitnote">
            <button onClick={() => {
              addNotetoText(props.elemId, props.selectedText, textRef.current.value)
              addNote();
            }}>
              submit
            </button>
          </div>
        )}
      </NoteContext.Consumer>
      <div className="eachnote">
        {childnotes.length && childnotes.map(note =>
          <React.Fragment>
            <div key={note}>
              <p>{Object.keys(note)[0]}</p>
              <p>{note[Object.keys(note)[0]]}</p>
            </div>
          </React.Fragment>
        )}
      </div>
    </div>
  )
});


class EachNotes extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      showmemorandum: true,
      selectedText: null,
      elemId: null,
      count: 0
    }
    this.detailRef = null;
    this.noteRef = React.createRef();
    this.titleRef = React.createRef();
    this.hintRef = React.createRef();
    this.makenoteRef = React.createRef();
    this.makenoteRef_list = React.createRef();
    this.makenoteRef_listdetail = React.createRef();

    this.isMemorandumFunc = this.isMemorandumFunc.bind(this);
    this.showmemorandumFunc = this.showmemorandumFunc.bind(this);

    this.addnewMemorandum = this.addnewMemorandum.bind(this);
    this.notesSelectText = this.notesSelectText.bind(this);
    this.makeNotetoText = this.makeNotetoText.bind(this);
  }
  componentDidUpdate() {
    if (!this.detailRef) {
      return;
    } else {
      this.detailRef.addEventListener('mouseup', this.notesSelectText);
    }
  }
  // deside if memorandum shows
  isMemorandumFunc(isShow) {
    this.setState({ showmemorandum: isShow });
  }

  // Click each memorandum to show its detail contents and close blank memorandum
  showmemorandumFunc(id) {
    this.setState({ elemId: id });
    this.props.showDetail(id);
    this.isMemorandumFunc(true);
    this.hintRef.current.classList.remove('active');
    this.makenoteRef.current.classList.remove('active');
    this.makenoteRef_list.current.classList.remove('active');
    this.makenoteRef_listdetail.current.classList.remove('active');
  }

  // Add new memorandum
  addnewMemorandum() {
    let name = this.props.name;
    let note = {
      'title': this.titleRef.current.value,
      'content': this.noteRef.current.value,
      'date': Date.now(),
      'id': `${name}_${this.props.notes.length}`,
      'textNote': []
    }
    this.props.addnewMemorandum(name, note)
  }

  // Select a paragraph in note
  notesSelectText(event) {
    // Get selected text and add tag "span" to it.
    let text = window.getSelection();
    let elem = document.createElement("span");
    elem.id = `${this.state.elemId}_${this.state.count}`;
    text.getRangeAt(0).surroundContents(elem);

    // Record selected text as string
    let selectedText = text.toString();
    this.setState({ selectedText: selectedText });

    // show hintRef (underline and addNotetoText)
    this.hintRef.current.classList.add('active');
    this.hintRef.current.style.left = `${event.clientX}px`;
    this.hintRef.current.style.top = `${event.clientY}px`;
  }

  // Make note to selected text
  makeNotetoText() {
    // Underline selected text
    document.getElementById(`${this.state.elemId}_${this.state.count}`).style.textDecoration = 'underline';
    this.setState({ count: this.state.count + 1 });
    // show makenoteRef area
    this.makenoteRef_list.current.classList.add('active');
    this.makenoteRef_listdetail.current.classList.add('active');
    this.makenoteRef.current.classList.add('active');
  }

  render() {
    const notes = this.props.notes;
    const name = this.props.name;
    return (
      <div className="container">
        <div className="row justify-content-center">
          <div ref={this.makenoteRef_list} className="col-6 list">
            <div data-testid="add_icon" onClick={this.isMemorandumFunc.bind(this, false)}>
              <i className="material-icons">add_circle_outline</i>
            </div>
            <div>
              {notes && notes.map((note, index) =>
                <div key={note.id} onClick={this.showmemorandumFunc.bind(this, `${name}_${index}`)}>
                  <p data-testid={note.id}>{note.title}</p>
                </div>
              )}
            </div>
          </div>
          <div ref={this.makenoteRef_listdetail} className="col-6 list_detail">
            {this.state.showmemorandum && notes && notes.map((note, index) => {
              if (note.textNote[0]) {
                let newnote1 = note.content.split(Object.keys(note.textNote[0]))[0];
                let newnote2 = Object.keys(note.textNote[0]);
                let newnote3 = note.content.split(Object.keys(note.textNote[0]))[1];
                return note.showdetail ?
                  <p ref={elem => this.detailRef = elem} key={`${note}-${index}`}>
                    {newnote1}
                    <span
                      style={{ textDecoration: 'underline' }}
                      onMouseOver={() => {
                        this.makenoteRef.current.classList.add('active');
                        this.makenoteRef_list.current.classList.add('active');
                        this.makenoteRef_listdetail.current.classList.add('active');
                      }}>
                      {newnote2}
                    </span>
                    {newnote3}
                  </p>
                  : null
              } else {
                return note.showdetail ?
                  <p ref={elem => this.detailRef = elem} key={index}>{note.content}</p>
                  : null
              }
            })}
            <div className="hint" ref={this.hintRef}>
              <div onClick={this.makeNotetoText}>Add Note</div>
            </div>
            {this.state.showmemorandum === false &&
              <div className="addmemorandum">
                <input data-testid="memorandum_title" ref={this.titleRef} />
                <textarea data-testid="memorandum_content" rows="18" ref={this.noteRef}></textarea>
                <div>
                  <button data-testid="memorandum_submit" onClick={this.addnewMemorandum}>Submit</button>
                </div>
              </div>
            }
          </div>
          <AddNotetoTextFunc
            ref={this.makenoteRef}
            name={name}
            elemId={this.state.elemId}
            selectedText={this.state.selectedText}
            hintRef={this.hintRef}
            notes={notes}
          />
        </div>
      </div>
    )
  }
}

const NoteContext = React.createContext();
class EachSubject extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      data: props.data,
      application: null,
      img: null,
      name: '',
    }
    this.chooseItem = this.chooseItem.bind(this);
  }
  chooseItem(eachitem) {
    this.setState({
      application: eachitem.application,
      img: eachitem.img,
      name: eachitem.name,
    });
    this.props.showSubject(eachitem.name);
  }
  render() {
    return (
      <React.Fragment>
        <section className="subject_list">
          <div className="container">
            <div className="row">
              <div className="col-12">
                <ul className="subjects">
                  {this.state.data.map((eachitem) => (
                    <li data-testid={`subject_${eachitem.name}`} className={`subject_${eachitem.name}`} onClick={() => this.chooseItem(eachitem)} key={eachitem.name}>
                      <span>{eachitem.name}</span>
                    </li>
                  ))}
                </ul>
              </div>
            </div>
          </div>
        </section>
        <section className="subject_detail">
          <div className="container">
            <div className="row justify-content-center">
              {this.props.render(this.state)}
            </div>
          </div>
        </section>
        <section className="subject_notes">
          <NoteContext.Provider value={{ addNotetoText: this.props.addNotetoText }}>
            <EachNotes
              notes={this.props.memorandum}
              showDetail={this.props.showDetail}
              name={this.state.name}
              addnewMemorandum={this.props.addnewMemorandum}
            />
          </NoteContext.Provider>
        </section>
      </React.Fragment>
    )
  }
}

class SubjectDetail extends React.Component {
  constructor(props) {
    super(props)
  }
  render() {
    const subdetail = this.props.subdetail;
    return (
      <React.Fragment>
        <div className="col-12 col-md-6 picture">
          <img data-testid="content_img" src={subdetail.img? subdetail.img : defaultImg} alt='img' className='inline-photo show-on-scroll' />
        </div>
        <div className="col-12 col-md-6 describe">
          <p data-testid="content" className="content">{subdetail.application}</p>
        </div>
      </React.Fragment>
    )
  }
}

class Subjects extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      data: props.data,
    }
  }
  componentDidMount() {
    // Detect request animation frame
    var scroll = window.requestAnimationFrame ||
      // IE Fallback
      function (callback) { window.setTimeout(callback, 1000 / 60) };

    var elementsToShow = document.querySelectorAll('.show-on-scroll');

    function loop() {
      Array.prototype.forEach.call(elementsToShow, function (element) {
        if (isElementInViewport(element)) {
          element.classList.add('is-visible');
        } else {
          element.classList.remove('is-visible');
        }
      });
      scroll(loop);
    }
    // Call the loop for the first time
    loop();

    // Helper function from: http://stackoverflow.com/a/7557433/274826
    function isElementInViewport(el) {
      // special bonus for those using jQuery
      if (typeof jQuery === "function" && el instanceof jQuery) {
        el = el[0];
      }
      var rect = el.getBoundingClientRect();
      return (
        (rect.top <= 0 && rect.bottom >= 0) ||
        (rect.bottom >= (window.innerHeight || document.documentElement.clientHeight) &&
          rect.top <= (window.innerHeight || document.documentElement.clientHeight)) ||
        (rect.top >= 0 && rect.bottom <= (window.innerHeight || document.documentElement.clientHeight))
      );
    }
  }
  render() {
    return (
      <React.Fragment>
        <EachSubject {...this.props} render={subdetail => (
          <SubjectDetail subdetail={subdetail} />
        )}
        />
      </React.Fragment>
    )
  }
}

function HigerOrderComponent(WrappedComponent, data) {
  return class extends React.Component {
    constructor(props) {
      super(props)
    }
    render() {
      return (
        <WrappedComponent data={data} {...this.props} />
      );
    }
  }
};

export const Business = HigerOrderComponent(Subjects, EachSubjectData.businessData);
export const Economics = HigerOrderComponent(Subjects, EachSubjectData.economicsData);


