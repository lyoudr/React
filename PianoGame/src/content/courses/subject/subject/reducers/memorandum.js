export const EachSubjectData = {
  businessData: {
    "Statistic": [
      { 'title': 'I write JavaScript today', 'content': 'You might be thinking that we’d need a separate effect to perform the cleanup. But code for adding and removing a subscription is so tightly related that useEffect is designed to keep it together. If your effect returns a function, React will run it when it is time to clean up', 'date': '2019.12.03', id: 'Statistic_0', textNote: [] },
      { 'title': 'Node JS', 'content': 'This is the optional cleanup mechanism for effects. Every effect may return a function that cleans up after it. This lets us keep the logic for adding and removing subscriptions close to each other. They’re part of the same effect!', 'date': '2019.11.20', id: 'Statistic_1', textNote: [] },
      { 'title': 'Angular', 'content': 'React performs the cleanup when the component unmounts. However, as we learned earlier, effects run for every render and not just once. This is why React also cleans up effects from the previous render before running the effects next time. ', 'date': '2019.11.05', id: 'Statistic_2', textNote: [] }
    ],
    "Commercial": [
      { 'title': 'Commercial', 'content': 'It is very good', 'date': '2019.12.03', id: 'Commercial_0', textNote: [] },
      { 'title': 'Node JS', 'content': 'It is very interesting', 'date': '2019.11.20', id: 'Commercial_1', textNote: [] },
      { 'title': 'Angular', 'content': 'It is very great', 'date': '2019.11.05', id: 'Commercial_2', textNote: [] }
    ],
    "Stock": [
      { 'title': 'business', 'content': 'It is very abc', 'date': '2019.12.03', id: 'Stock_0', textNote: [] },
      { 'title': 'Node JS', 'content': 'It is very des', 'date': '2019.11.20', id: 'Stock_1', textNote: [] },
      { 'title': 'Angular', 'content': 'It is very efsef', 'date': '2019.11.05', id: 'Stock_2', textNote: [] }
    ],
    "Fishery": [
      { 'title': 'business', 'content': 'It is very abc', 'date': '2019.12.03', id: 'Stock_0', textNote: [] },
      { 'title': 'Node JS', 'content': 'It is very des', 'date': '2019.11.20', id: 'Stock_1', textNote: [] },
      { 'title': 'Angular', 'content': 'It is very efsef', 'date': '2019.11.05', id: 'Stock_2', textNote: [] }
    ],
    "Geology": [
      { 'title': 'business', 'content': 'It is very abc', 'date': '2019.12.03', id: 'Stock_0', textNote: [] },
      { 'title': 'Node JS', 'content': 'It is very des', 'date': '2019.11.20', id: 'Stock_1', textNote: [] },
      { 'title': 'Angular', 'content': 'It is very efsef', 'date': '2019.11.05', id: 'Stock_2', textNote: [] }
    ],
    "legal": [
      { 'title': 'business', 'content': 'It is very abc', 'date': '2019.12.03', id: 'Stock_0', textNote: [] },
      { 'title': 'Node JS', 'content': 'It is very des', 'date': '2019.11.20', id: 'Stock_1', textNote: [] },
      { 'title': 'Angular', 'content': 'It is very efsef', 'date': '2019.11.05', id: 'Stock_2', textNote: [] }
    ]
  },
  economicsData: {
    "Statistic": [
      { 'title': 'I write JavaScript today', 'content': 'It is very interesting', 'date': '2019.12.03' },
      { 'title': 'Node JS', 'content': 'It is very interesting', 'date': '2019.11.20' },
      { 'title': 'Angular', 'content': 'It is very interesting', 'date': '2019.11.05' }
    ],
    "Commercial": [
      { 'title': 'Commercial', 'content': 'It is very interesting', 'date': '2019.12.03' },
      { 'title': 'business', 'content': 'It is very interesting', 'date': '2019.11.20' },
      { 'title': 'Angular', 'content': 'It is very interesting', 'date': '2019.11.05', id: '3' }
    ],
    "Stock": [
      { 'title': 'stock', 'content': 'It is very interesting', 'date': '2019.12.03' },
      { 'title': 'Node JS', 'content': 'It is very interesting', 'date': '2019.11.20' },
      { 'title': 'Angular', 'content': 'It is very interesting', 'date': '2019.11.05', id: '3' }
    ]
  }
}

export const memorandum = (state = EachSubjectData.businessData['Statistic'], action) => {
  switch (action.type) {
    case 'SHOW_SUBJECT':
      return EachSubjectData.businessData[action.name]
    case 'ADD_MEMO':
      return [
        ...EachSubjectData.businessData[action.name],
        action.note
      ]
    case 'SHOW_DETAIL':
      return state.map(memorandum => {
        if (memorandum.id === action.id) {
          memorandum.showdetail = true;
          return memorandum;
        } else {
          memorandum.showdetail = false;
          return memorandum;
        }
      });
    case 'ADD_NOTETOTEXT':
      let subject = action.id.split('_')[0];
      EachSubjectData.businessData[subject].forEach(item => {
        if (item.id === action.id) {
          let newnote = {};
          newnote[action.text] = action.note;
          item.textNote.push(newnote);
        }
      });
      return EachSubjectData.businessData[subject]
    default:
      return state
  }
}


