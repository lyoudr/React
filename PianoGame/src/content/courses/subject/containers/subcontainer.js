import { connect } from 'react-redux';
import { showSubject } from '../../../../redux/actions';
import { Business, Economics} from '../../Subject';

const mapStateToProps = state => ({
    memorandum: state.memorandum
});

const mapDispatchToProps = dispatch => ({
    showSubject: name => dispatch(showSubject(name)),
});

export const Business = connect(
    mapStateToProps,
    mapDispatchToProps
)(Business)

export const Economics = connect(
    mapStateToProps,
    mapDispatchToProps
)(Economics)