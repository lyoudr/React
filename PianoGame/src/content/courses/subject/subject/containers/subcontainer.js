import { connect } from 'react-redux';
import { showSubject, showDetail, addnewMemorandum, addNotetoText } from '../../../../../redux/actions';
import { Business, Economics} from '../presentcomp/Subject';

const mapStateToProps = state => ({
    memorandum: state.memorandum
});

const mapDispatchToProps = dispatch => ({
    showSubject: name => {
        dispatch(showSubject(name))
    },
    showDetail : id => {
        dispatch(showDetail(id));
    },
    addnewMemorandum : (name, note) => {
        dispatch(addnewMemorandum(name, note));
    },
    addNotetoText : (id, text, note) => {
        dispatch(addNotetoText(id, text, note));
    }
});

export const BusinessContainer = connect(
    mapStateToProps,
    mapDispatchToProps
)(Business)

export const EconomicsContainer = connect(
    mapStateToProps,
    mapDispatchToProps
)(Economics)