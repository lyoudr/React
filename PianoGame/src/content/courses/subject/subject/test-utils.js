import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { createStore } from 'redux';
import { EachSubjectData as reducerInitialState, memorandum } from './reducers/memorandum';
import { Provider } from 'react-redux';

function render(ui, {
  initialState = reducerInitialState.businessData['Statistic'],
  store = createStore(memorandum, initialState),
  ...renderOptions
} = {}) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';

// override render method
export { render };