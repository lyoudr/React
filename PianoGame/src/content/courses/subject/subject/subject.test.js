import React from 'react';
// We're using our own custom render function and not RTL's render
// our customer utils also re-export everything from RTL 
// so we can import fireEvent and screen here as well
import renderer from 'react-test-renderer';
import { render, fireEvent, screen } from './test-utils';
import { Business } from './presentcomp/Subject';
import { mapStateToProps, mapDispatchToProps } from './containers/subcontainer';
import { EachSubjectData, memorandum } from './reducers/memorandum';
import { createStore } from 'redux';

describe('Subject test', () => {
  const state = {
    memorandum: EachSubjectData.businessData['Statistic']
  };
  // store
  let store;
  // state and funciton
  let memorandum_data;
  let showSubject;
  let addnewMemorandum;
  
  beforeEach(() => {
    store = createStore(memorandum, state.memorandum),
    memorandum_data = mapStateToProps(state).memorandum;
    showSubject = (name) => mapDispatchToProps(store.dispatch).showSubject(name);
    addnewMemorandum = (name, note) => mapDispatchToProps(store.dispatch).addnewMemorandum(name, note);
  });
  
  it('renders correctly', () => {
    const component = renderer.create(<Business />);
    //1. rendering
    let json_tree = component.toJSON();
    expect(json_tree).toMatchSnapshot();
  });

  test('can render with redux with defaults', () => {
    const business = render(<Business showSubject = {showSubject}/>);
    const statistic_btn = screen.getByTestId('subject_Statistic');
    const content_img = screen.getByTestId('content_img');
    const content = screen.getByTestId('content');
    fireEvent.click(statistic_btn);
    expect(content_img.src).toMatch("http://localhost/business1.jpg")
    expect(content.textContent).toContain('statistic');
  });

  test('should add right', async() => {
    const {business, rerender} = render(
    <Business 
      showSubject = {showSubject} 
      addnewMemorandum = {addnewMemorandum}
      memorandum = {memorandum_data}
    />);
    // choose subject
    const statistic_btn = screen.getByTestId('subject_Statistic');
    fireEvent.click(statistic_btn);
    // click add note button
    const add_btn = screen.getByTestId('add_icon');
    fireEvent.click(add_btn);
    // expect add note area to show
    const memorandum_title = screen.getByTestId('memorandum_title');
    const memorandum_content = screen.getByTestId('memorandum_content');
    expect(memorandum_title).toBeTruthy();
    expect(memorandum_content).toBeTruthy();
    // add note right
    fireEvent.change(memorandum_title, { target: { value: 'Piano' }});
    fireEvent.change(memorandum_content, {target: { value: 'This is a note about piano.'}});
    expect(memorandum_title.value).toMatch(/Piano/);
    expect(memorandum_content.value).toContain('This is a note about piano.');  
    // submit notes
    const memorandum_submit = screen.getByTestId('memorandum_submit');
    fireEvent.click(memorandum_submit);
    // re-render the same component by using new state from "getState()"
    rerender(<Business 
      showSubject = {showSubject} 
      addnewMemorandum = {addnewMemorandum}
      memorandum = {store.getState()}
    />)
    const new_note = screen.getByTestId('Statistic_3');
    expect(new_note.textContent).toMatch(/Piano/);
  });
});