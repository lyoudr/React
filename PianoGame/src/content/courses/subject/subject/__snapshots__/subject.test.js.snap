// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subject test renders correctly 1`] = `
Array [
  <section
    className="subject_list"
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <div
          className="col-12"
        >
          <ul
            className="subjects"
          >
            <li
              className="subject_Statistic"
              data-testid="subject_Statistic"
              onClick={[Function]}
            >
              <span>
                Statistic
              </span>
            </li>
            <li
              className="subject_Commercial"
              data-testid="subject_Commercial"
              onClick={[Function]}
            >
              <span>
                Commercial
              </span>
            </li>
            <li
              className="subject_Stock"
              data-testid="subject_Stock"
              onClick={[Function]}
            >
              <span>
                Stock
              </span>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </section>,
  <section
    className="subject_detail"
  >
    <div
      className="container"
    >
      <div
        className="row justify-content-center"
      >
        <div
          className="col-12 col-md-6 picture"
        >
          <img
            alt="img"
            className="inline-photo show-on-scroll"
            data-testid="content_img"
            src="subject_default.jpeg"
          />
        </div>
        <div
          className="col-12 col-md-6 describe"
        >
          <p
            className="content"
            data-testid="content"
          />
        </div>
      </div>
    </div>
  </section>,
  <section
    className="subject_notes"
  >
    <div
      className="container"
    >
      <div
        className="row justify-content-center"
      >
        <div
          className="col-6 list"
        >
          <div
            data-testid="add_icon"
            onClick={[Function]}
          >
            <i
              className="material-icons"
            >
              add_circle_outline
            </i>
          </div>
          <div />
        </div>
        <div
          className="col-6 list_detail"
        >
          <div
            className="hint"
          >
            <div
              onClick={[Function]}
            >
              Add Note
            </div>
          </div>
        </div>
        <div
          className="addnotetoText"
        >
          <textarea />
          <div
            className="submitnote"
          >
            <button
              onClick={[Function]}
            >
              submit
            </button>
          </div>
          <div
            className="eachnote"
          >
            0
          </div>
        </div>
      </div>
    </div>
  </section>,
]
`;
